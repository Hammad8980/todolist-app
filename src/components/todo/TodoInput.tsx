import { useState, type FormEvent } from 'react';
import Button from '../ui/Button';
type TodoInputProps = {
  onAddTask: (taskName: string) => void;
};

function TodoInput({ onAddTask }: TodoInputProps) {
  const [taskName, setTaskName] = useState('');
  const handleAddTasks = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault(); // Sometimes page is refreshed upon trigger of Event so this property is to prevent the browser from doing that
    onAddTask(taskName);
    setTaskName('');
  };
  return (
    <form className="flex" onSubmit={handleAddTasks}>
      {/*Below is the traditional onKeyDown method Commented for learning purposes*/}
      {/* <input
          value={taskName}
          placeholder='Enter text...'
          onChange={(e) => setTaskName(e.target.value)}
          onKeyDown={(e)=> { if (e.key === "Enter") { handleAddTasks(); }}} // Enter key is send
        />
        <button
          className="bg-blue-500 hover:bg-blue-700 text-white font-bold ml-2 pb-0.5 px-2 rounded"
          onClick={handleAddTasks}
        >
          +
        </button> */}
      <input
        type="text"
        value={taskName}
        placeholder="Enter text..."
        onChange={e => setTaskName(e.target.value)}
        className="flex-grow p-2 border rounded-l focus:outline-none focus:ring-2 focus:ring-blue-300"
      />
      <Button
        className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 ml-2 rounded-r rounded-l"
        type="submit"
      >
        +
      </Button>
    </form>
  );
}

export default TodoInput;
