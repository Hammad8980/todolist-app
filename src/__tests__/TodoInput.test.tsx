import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TodoInput from '../components/todo/TodoInput';

describe('TodoInput', () => {
  const mockAddTask = jest.fn();

  beforeEach(() => {
    mockAddTask.mockClear();
  });

  it('renders input field and button correctly', () => {
    const { container } = render(<TodoInput onAddTask={mockAddTask} />);

    const input = screen.getByPlaceholderText(/enter text/i); // Using a regex to match the placeholder text
    const button = screen.getByRole('button', { name: '+' });

    expect(input).toBeInTheDocument();
    expect(button).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('updates input value when typing', async () => {
    const user = userEvent.setup();
    const { container } = render(<TodoInput onAddTask={mockAddTask} />);

    const input = screen.getByPlaceholderText(/enter text/i) as HTMLInputElement; // Using a type assertion to specify the input type

    await user.type(input, 'New task');

    expect(input.value).toBe('New task');
    expect(container).toMatchSnapshot();
  });

  it('calls onAddTask with correct value when form is submitted via button click', async () => {
    const user = userEvent.setup();
    const { container } = render(<TodoInput onAddTask={mockAddTask} />);

    const input = screen.getByPlaceholderText(/enter text/i);
    const button = screen.getByRole('button', { name: '+' });

    await user.type(input, 'Test Task');
    await user.click(button);

    expect(mockAddTask).toHaveBeenCalledWith('Test Task');
    expect(mockAddTask).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot();
  });

  it('calls onAddTask when form is submitted via Enter key', async () => {
    const user = userEvent.setup();
    const { container } = render(<TodoInput onAddTask={mockAddTask} />);

    const input = screen.getByPlaceholderText(/enter text/i);

    await user.type(input, 'Test Task{enter}');

    expect(mockAddTask).toHaveBeenCalledWith('Test Task');
    expect(mockAddTask).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot();
  });

  it('clears input field after successful submission', async () => {
    const user = userEvent.setup(); //
    const { container } = render(<TodoInput onAddTask={mockAddTask} />);

    const input = screen.getByPlaceholderText(/enter text/i) as HTMLInputElement;
    const button = screen.getByRole('button', { name: '+' });

    await user.type(input, 'Test Task');
    await user.click(button);

    expect(input.value).toBe('');
    expect(container).toMatchSnapshot();
  });

  it('does not call onAddTask when input is empty', async () => {
    const user = userEvent.setup();
    const { container } = render(<TodoInput onAddTask={mockAddTask} />);

    const button = screen.getByRole('button', { name: '+' });

    await user.click(button);

    expect(mockAddTask).not.toHaveBeenCalledWith();
    expect(container).toMatchSnapshot();
  });

  it('handles form submission correctly', () => {
    const { container } = render(<TodoInput onAddTask={mockAddTask} />);

    const input = screen.getByPlaceholderText(/enter text/i);
    const form = input.closest('form');

    fireEvent.change(input, { target: { value: 'My Task' } });
    fireEvent.submit(form!);

    expect(mockAddTask).toHaveBeenCalledWith('My Task');
    expect(mockAddTask).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot();
  });
});
