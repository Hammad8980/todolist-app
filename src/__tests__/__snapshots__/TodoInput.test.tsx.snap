// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoInput calls onAddTask when form is submitted via Enter key 1`] = `
<div>
  <form
    class="flex"
  >
    <input
      class="flex-grow p-2 border rounded-l focus:outline-none focus:ring-2 focus:ring-blue-300"
      placeholder="Enter text..."
      type="text"
      value=""
    />
    <button
      class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 ml-2 rounded-r rounded-l"
      type="submit"
    >
      +
    </button>
  </form>
</div>
`;

exports[`TodoInput calls onAddTask with correct value when form is submitted via button click 1`] = `
<div>
  <form
    class="flex"
  >
    <input
      class="flex-grow p-2 border rounded-l focus:outline-none focus:ring-2 focus:ring-blue-300"
      placeholder="Enter text..."
      type="text"
      value=""
    />
    <button
      class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 ml-2 rounded-r rounded-l"
      type="submit"
    >
      +
    </button>
  </form>
</div>
`;

exports[`TodoInput clears input field after successful submission 1`] = `
<div>
  <form
    class="flex"
  >
    <input
      class="flex-grow p-2 border rounded-l focus:outline-none focus:ring-2 focus:ring-blue-300"
      placeholder="Enter text..."
      type="text"
      value=""
    />
    <button
      class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 ml-2 rounded-r rounded-l"
      type="submit"
    >
      +
    </button>
  </form>
</div>
`;

exports[`TodoInput does not call onAddTask when input is empty 1`] = `
<div>
  <form
    class="flex"
  >
    <input
      class="flex-grow p-2 border rounded-l focus:outline-none focus:ring-2 focus:ring-blue-300"
      placeholder="Enter text..."
      type="text"
      value=""
    />
    <button
      class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 ml-2 rounded-r rounded-l"
      type="submit"
    >
      +
    </button>
  </form>
</div>
`;

exports[`TodoInput handles form submission correctly 1`] = `
<div>
  <form
    class="flex"
  >
    <input
      class="flex-grow p-2 border rounded-l focus:outline-none focus:ring-2 focus:ring-blue-300"
      placeholder="Enter text..."
      type="text"
      value=""
    />
    <button
      class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 ml-2 rounded-r rounded-l"
      type="submit"
    >
      +
    </button>
  </form>
</div>
`;

exports[`TodoInput renders input field and button correctly 1`] = `
<div>
  <form
    class="flex"
  >
    <input
      class="flex-grow p-2 border rounded-l focus:outline-none focus:ring-2 focus:ring-blue-300"
      placeholder="Enter text..."
      type="text"
      value=""
    />
    <button
      class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 ml-2 rounded-r rounded-l"
      type="submit"
    >
      +
    </button>
  </form>
</div>
`;

exports[`TodoInput updates input value when typing 1`] = `
<div>
  <form
    class="flex"
  >
    <input
      class="flex-grow p-2 border rounded-l focus:outline-none focus:ring-2 focus:ring-blue-300"
      placeholder="Enter text..."
      type="text"
      value="New task"
    />
    <button
      class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 ml-2 rounded-r rounded-l"
      type="submit"
    >
      +
    </button>
  </form>
</div>
`;
